File: internal/domain/grid.go
package domain

// GridRequest is the POST body for server-side grid operations
type GridRequest struct {
	StartRow    int                   `json:"startRow"`
	EndRow      int                   `json:"endRow"`
	SortModel   []SortItem            `json:"sortModel"`
	FilterModel map[string]FilterItem `json:"filterModel"`
}

type SortItem struct {
	ColID string `json:"colId"`
	Sort  string `json:"sort"` // "asc" or "desc"
}

type FilterItem struct {
	FilterType string      `json:"filterType"` // "text", "number", "date"
	Type       string      `json:"type"`       // "contains", "equals", "startsWith", "endsWith", "inRange", "set"
	Filter     interface{} `json:"filter"`
	FilterTo   interface{} `json:"filterTo"` // for inRange
	Values     []string    `json:"values"`   // for set filter
}

// GridResponse is the standard response for grid queries
type GridResponse struct {
	RowsData      interface{} `json:"rowsData"`
	TotalRowCount int64       `json:"totalRowCount"`
}

// BulkUpdateRequest for async bulk operations
type BulkUpdateRequest struct {
	Updates []BulkUpdateItem `json:"updates"`
}

type BulkUpdateItem struct {
	MaHang string                 `json:"ma_hang"`
	Fields map[string]interface{} `json:"fields"`
}

// BulkUpdateResponse returns job ID for polling
type BulkUpdateResponse struct {
	JobID string `json:"job_id"`
}

File: web/src/types/grid.ts
export interface SortItem {
	colId: string;
	sort: "asc" | "desc";
}

export interface FilterItem {
	filterType: "text" | "number" | "date";
	type: "contains" | "equals" | "startsWith" | "endsWith" | "inRange" | "set";
	filter?: string | number;
	filterTo?: string | number;
	values?: string[];
}

export interface GridRequest {
	startRow: number;
	endRow: number;
	sortModel: SortItem[];
	filterModel: Record<string, FilterItem>;
}

export interface GridResponse<T = Record<string, unknown>> {
	rowsData: T[];
	totalRowCount: number;
}

export interface BulkUpdateItem {
	ma_hang: string;
	fields: Record<string, unknown>;
}

export interface BulkUpdateRequest {
	updates: BulkUpdateItem[];
}

export interface BulkUpdateResponse {
	job_id: string;
}

export interface AsyncJob {
	job_id: string;
	job_type: string;
	status: "pending" | "running" | "completed" | "failed";
	payload: string;
	result: string;
	error: string;
	created_at: string;
	updated_at: string;
}
